### PR TITLE
chat: added oembed parser

### PR DESCRIPTION
**Pull request by tylershuster**
This PR adds oembed parsing to chat. OEmbedding is an open protocol to specify embedding content in pages from a URL. Examples include YouTube and Twitter.  Here's a video of it working:

[Video](https://marpem-files.sfo2.digitaloceanspaces.com/Screen%20Recording%202020-08-27%20at%203.08.04%20PM.mov)

Sick, right?

## Problems This Fixes
- There's currently an exception for youtube to iframe its content, which is Weird given urbit's position toward MEGACORP. This implements an open protocol that youtube *also* implements. Incidentally their oembed response is just the iframe that chat already uses.
- Obviously, it vastly expands the possibilities for oembeds
- It replaces the links of oembeddable content with their title, if available. I had someone ask me about doing this via the api and ew.

Ok here's some

## Caveats
- It uses https://noembed.com/ to bypass the CORS restrictions that would happen with a vanilla implementation
  - Counter-caveat a: This could be avoided by proxing requests through arvo.
  - Counter-caveat b: A *lot* of people use this service. Assuming the above solution is eventually implemented, this wouldn't result in a significantly diminished experience if noembed were to go down.
- It makes a request to a third party without explicit permission
  - Counter-caveat: It already does this; see: images.

## Future Work
I've always imagined this feature being paired with a new section in settings called "Remote Content."

```
[ ] Load images
[ ] Load videos
[ ] Load audio
[ ] Load oembeds
```

And we default to images being on, I guess, because that's what people are used to. The idea though is that your browser shouldn't load any domains except the one you're on.

Anyway I see these two features as going hand-in-hand.

RFC

### CODE REVIEW

**Comment by tacryt-socryp** [comment]
I prefer this approach to remote content to the current one. Looks excellent.

**Comment by tylershuster** [comment]
Reviewing, I think we probably don't want to render any remote content (even under `dn`) without clicking embed.

**Comment by matildepark** [comment]
Might be worth componentising this somehow. Links uses similar stuff.

**Comment by matildepark** [comment]
Are you intending to revise the PR or submit subsequent changes separately?

**Comment by tylershuster** [comment]
Revising today or tomorrow. I’ve been outside. 

T

**Comment by matildepark** [comment]
That’s what I like to hear!

—
~haddef-sigwen
https://urbit.org

**Comment by tylershuster** [comment]
This is my first pass at a new local setting. Seems to work, unimplemented as of yet.

Next step is to incorporate these new settings into a) `views/apps/chat/compontents/lib/content/url.js` and b) `views/apps/links/components/lib/link-detail-preview.js`

Comments appreciated

**Comment by tylershuster** [comment]
Alright, there's actually something to look at here. Some styling needs to be included, but I think the basic logic is solid. Defaults to loading everything except oembeds.

**Comment by matildepark** [comment]
Will review in the morning.

File: pkg/interface/src/views/apps/profile/components/lib/RemoteContent.tsx [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+              <Checkbox
+                label="Load Audio"
+                id="audioShown"
+              />
+              <Checkbox
+                label="Load Videos"
+                id="videoShown"
+              />
+              <Checkbox
+                label="Load Oembeds"
```
Comment (matildepark) [inline pkg/interface/src/views/apps/profile/components/lib/RemoteContent.tsx:81]:
```suggestion
                label="Load embedded content"
```

File: pkg/interface/src/views/apps/profile/components/lib/RemoteContent.tsx [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+              <Checkbox
+                label="Load Images"
+                id="imageShown"
+              />
+              <Checkbox
+                label="Load Audio"
+                id="audioShown"
+              />
+              <Checkbox
+                label="Load Videos"
```
Comment (matildepark) [inline pkg/interface/src/views/apps/profile/components/lib/RemoteContent.tsx:77]:
```suggestion
                label="Load video files"
```

(Do we even have this?)

Comment (matildepark) [reply pkg/interface/src/views/apps/profile/components/lib/RemoteContent.tsx:77]: So we do.

File: pkg/interface/src/views/apps/profile/components/lib/RemoteContent.tsx [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+            <Box color="black" fontSize={1} mb={3} fontWeight={900}>
+              Remote Content Policy
+            </Box>
+            <Box>
+              <Checkbox
+                label="Load Images"
+                id="imageShown"
+              />
+              <Checkbox
+                label="Load Audio"
```
Comment (matildepark) [inline pkg/interface/src/views/apps/profile/components/lib/RemoteContent.tsx:73]:
```suggestion
                label="Load audio files"
```

File: pkg/interface/src/views/apps/profile/components/lib/RemoteContent.tsx [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+            gridTemplateColumns="1fr"
+            gridTemplateRows="audio"
+            gridRowGap={3}
+          >
+            <Box color="black" fontSize={1} mb={3} fontWeight={900}>
+              Remote Content Policy
+            </Box>
+            <Box>
+              <Checkbox
+                label="Load Images"
```
Comment (matildepark) [inline pkg/interface/src/views/apps/profile/components/lib/RemoteContent.tsx:69]: Load images

File: pkg/interface/src/views/apps/profile/components/lib/RemoteContent.tsx [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+    >
+      {(props) => (
+        <Form>
+          <Box
+            display="grid"
+            gridTemplateColumns="1fr"
+            gridTemplateRows="audio"
+            gridRowGap={3}
+          >
+            <Box color="black" fontSize={1} mb={3} fontWeight={900}>
```
Comment (matildepark) [inline pkg/interface/src/views/apps/profile/components/lib/RemoteContent.tsx:64]: You should do font stuff in a separate component, `<Text>`.

File: pkg/interface/src/views/apps/profile/components/lib/RemoteContent.tsx [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+              Remote Content Policy
+            </Box>
+            <Box>
```
Comment (matildepark) [inline pkg/interface/src/views/apps/profile/components/lib/RemoteContent.tsx:67]: You'll want a `<Text gray>` that says "Embed remote content within this Landscape session."

File: pkg/interface/src/views/apps/profile/components/lib/RemoteContent.tsx [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+      {(props) => (
+        <Form>
+          <Box
+            display="grid"
+            gridTemplateColumns="1fr"
+            gridTemplateRows="audio"
+            gridRowGap={3}
+          >
+            <Box color="black" fontSize={1} mb={3} fontWeight={900}>
+              Remote Content Policy
```
Comment (matildepark) [inline pkg/interface/src/views/apps/profile/components/lib/RemoteContent.tsx:65]: "Remote Content" is fine.

File: pkg/interface/src/views/apps/profile/components/lib/RemoteContent.tsx (hunk 5 above)
Comment (tylershuster) [reply pkg/interface/src/views/apps/profile/components/lib/RemoteContent.tsx:64]: Tell that to whomever wrote `DisplayForm.tsx` ah ah ah

Comment (tylershuster) [reply pkg/interface/src/views/apps/profile/components/lib/RemoteContent.tsx:64]: Really though, what should I use for `Text` settings? All the other sections on this page use `Box`

**Comment by tylershuster** [comment]
Ok, this is good feedback. I added a few options to the `<RemoteContent>` component that increase its contextability (contextualizationability?). Then I applied those in chat and in links, to more or less maintain the existing behavior with new behavior added on top of it.

**Comment by matildepark** [comment]
Screenshot 2020-09-02 at 5 18 40 PM

YouTube embeds now cause horizontal overflow on mobile.

Screenshot 2020-09-02 at 5 19 10 PM

Links should not have the 'collapse' button available at all, basically.

**Comment by matildepark** [comment]
How did this conflict?

**Comment by tylershuster** [comment]
Some jerk pushed a change to links earlier today. something to do with s3 i don't know

**Comment by matildepark** [comment]
ok will test, just clean some of the commits here and we'll get going probably

**Review by matildepark — APPROVED** [review]
Feels very nice. Some slight padding issues with the settings console here, but we're probably just doing a holistic pass later anyway.

**Comment by tylershuster** [comment]
Can you give this a second look? I've had some merging issues this evening.

File: pkg/interface/src/views/apps/profile/components/lib/RemoteContent.tsx (hunk 1 above)
Comment (matildepark) [inline pkg/interface/src/views/apps/profile/components/lib/RemoteContent.tsx:81]: We lost these changes.

**Comment by matildepark** [comment]
That seems to be the only regression, settings copy.

**Comment by tylershuster** [comment]
Ok, that was kind of weird, but as long as it works now.

File: pkg/interface/src/views/apps/profile/components/lib/RemoteContent.tsx [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+              <Checkbox
+                label="Load images"
+                id="imageShown"
+              />
+              <Checkbox
+                label="Load audio files"
+                id="audioShown"
+              />
+              <Checkbox
+                label="Load field files"
```
Comment (matildepark) [inline pkg/interface/src/views/apps/profile/components/lib/RemoteContent.tsx:77]: What are field files ...

Comment (tylershuster) [reply pkg/interface/src/views/apps/profile/components/lib/RemoteContent.tsx:77]: ihaventhadmycoffee.jpg

**Pull request merged**